### PR TITLE
std.crypto: add DER parser and RSA

### PR DESCRIPTION
**Pull request by clickingbuttons**
- Add secure DER parser.
	- Use in ECDSA.
		- Add comments to Wycheproof tests.
		- Move Wycheproof test suite to external file.
	- Use in Ed25519.
	- Not public.
	- Ready for use in Certificate, which will be done in another PR for ease of review.
- Add std.crypto.rsa.
	- Modulus_bits up to 4096.
	- Any hash function for PSS or PKCS1v1.5 signatures.
	- Use in Certificate and tls.Client.

### CODE REVIEW

File: lib/std/crypto/rsa.zig [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
+                const em = em_buf[0..pk.modulus.byteLen()];
+                try emm.toBytes(em, .big);
+
+                var hash: [Hash.digest_length]u8 = undefined;
+                self.h.final(&hash);
+
+                // TODO: compare hash values instead of emsa values
+                const expected = try emsaEncode(hash, em);
+
+                if (!std.mem.eql(u8, expected, em)) return error.Inconsistent;
```
Comment (clickingbuttons) [inline lib/std/crypto/rsa.zig:400]: Not sure if this needs to be constant time or not. It's not mentioned anywhere, so I guessed it's okay to use the fast version.

File: lib/std/crypto/rsa.zig [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+        if (std.debug.runtime_safety) {
+            const p = try Fe.fromBytes(public.modulus, parser.view(prime1), .big);
+            const q = try Fe.fromBytes(public.modulus, parser.view(prime2), .big);
+
+            // check that n = p * q
+            const expected_zero = public.modulus.mul(p, q);
+            if (!expected_zero.isZero()) return error.KeyMismatch;
+
+            // TODO: check that d * e is one mod p-1 and mod q-1. Note d and e were bound
+            // const de = secret.private_exponent.mul(public.public_exponent);
```
Comment (clickingbuttons) [inline lib/std/crypto/rsa.zig:199]: Need some help here. Can also just remove this incorrect commented code for now.

File: lib/std/crypto/rsa.zig [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+            // just use the empty hash...
+            else => {},
+        }
+    }
+    var res: [Hash.digest_length]u8 = undefined;
+    Hash.hash(label, &res, .{});
+    return res;
+}
+
+const ct = if (std.options.side_channels_mitigations == .none) ct_unprotected else ct_protected;
```
Comment (clickingbuttons) [inline lib/std/crypto/rsa.zig:708]: Not sure what `ct` stands for, but I copied this pattern from `ff.zig`.

Comment (clickingbuttons) [reply lib/std/crypto/rsa.zig:708]: It stands for `constant_time`.

**Comment by jedisct1** [comment]
Once again, that should be split into more focused PRs.

Maybe starting with a PR that introduces a `crypto.serialization.asn1` namespace that exposes generic DER/PEM parsing (and plan for generation as well) and OIDs. 

`crypto.serialization` can include other codecs, especially constant-time base64 codecs that are required for PEM-encoded secrets.

And a PR that extracts the current RSA code from `Certificate.zig` and move it to a dedicated file. Without completely rewriting it, not the rest at that point. Then, incremental improvements can be made to it in order to address known compatibility issues with certificates chains.  

Once the serialization toolkit is done, the DER and PEM parser can be used everywhere, from `ecdsa` to `Bundle.zig`. Then, Wycheproof integration can be added.

Then, other improvements to `rsa.zig` can be made. In order to support blinding and operations on private key, we need to add things such as modular inversion to `ff`. So, individual PRs to implement these prerequisites. Then we need to make sure we can serialize to DER/PEM. Then we'll have everything to replace `rsa.zig` with a complete implementation, without question marks.

**Comment by clickingbuttons** [comment]
Sure, let's be surgical. I've split this up into issues:
- #19775
- #19776

**Pull request closed without merging**